lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -596,17 +596,62 @@ std::string SerializedModuleBaseName::getName(file_types::ID fileTy) const {
   return std::string(result.str());
 }
 
+llvm::Optional<std::string> SerializedModuleBaseName::getPackageInterfacePathIfInSamePackage(llvm::vfs::FileSystem &fs, ASTContext &ctx) const {
+
+  if (!ctx.LangOpts.EnablePackageInterfaceLoad)
+    return {};
+
+  std::string packagePath{
+    getName(file_types::TY_PackageSwiftModuleInterfaceFile)};
+
+  if (fs.exists(packagePath)) {
+    // Read the interface file and extract its package-name argument value
+    StringRef result;
+    if (auto packageFile = llvm::MemoryBuffer::getFile(packagePath)) {
+      llvm::BumpPtrAllocator alloc;
+      llvm::StringSaver argSaver(alloc);
+      SmallVector<const char*, 8> args;
+      (void)extractCompilerFlagsFromInterface(packagePath,
+                                              (*packageFile)->getBuffer(), argSaver, args);
+      for (unsigned I = 0, N = args.size(); I + 1 < N; I++) {
+        StringRef current(args[I]), next(args[I + 1]);
+        if (current == "-package-name") {
+          // Instead of `break` here, continue to get the last value in case of dupes,
+          // to be consistent with the default parsing logic.
+          result = next;
+        }
+      }
+    }
+    // Return the .package.swiftinterface path if the package name applies to
+    // the importer module.
+    if (!result.empty() && result == ctx.LangOpts.PackageName)
+      return packagePath;
+  }
+  return {};
+}
+
 llvm::Optional<std::string>
-SerializedModuleBaseName::findInterfacePath(llvm::vfs::FileSystem &fs) const {
+SerializedModuleBaseName::findInterfacePath(llvm::vfs::FileSystem &fs, ASTContext &ctx) const {
   std::string interfacePath{getName(file_types::TY_SwiftModuleInterfaceFile)};
+  // Ensure the public swiftinterface already exists, otherwise bail early
+  // as it's considered the module doesn't exist.
   if (!fs.exists(interfacePath))
     return {};
 
+  // Check if a package interface exists and if the package name applies to
+  // the importer module.
+  auto pkgPath = getPackageInterfacePathIfInSamePackage(fs, ctx).value_or("");
+  if (!pkgPath.empty() && fs.exists(pkgPath))
```

Comment:
Isn't the `fs.exists(pkgPath)` check redundant as `getPackageInterfacePathIfInSamePackage` returns only existing paths already?

Reply:
Apparently it doesn't! The fs method is in llvm so will need to fix it there.